src/batch/src/monitor/stats.rs

```diff
@@ -0,0 +1,418 @@
+
+        let task_idle_duration = GaugeVec::new(
+            opts!(
+                "batch_task_idle_duration",
+                "The total duration (s) that tasks idled.",
+            ),
+            &task_labels[..],
+        )
+        .unwrap();
+        descs.extend(task_idle_duration.desc().into_iter().cloned());
+
+        let task_poll_duration = GaugeVec::new(
+            opts!(
+                "batch_task_poll_duration",
+                "The total duration (s) elapsed during polls.",
+            ),
+            &task_labels[..],
+        )
+        .unwrap();
+        descs.extend(task_poll_duration.desc().into_iter().cloned());
+
+        let task_scheduled_duration = GaugeVec::new(
+            opts!(
+                "batch_task_scheduled_duration",
+                "The total duration (s) that tasks spent waiting to be polled after awakening.",
+            ),
+            &task_labels[..],
+        )
+        .unwrap();
+        descs.extend(task_scheduled_duration.desc().into_iter().cloned());
+
+        let task_slow_poll_duration = GaugeVec::new(
+            opts!(
+                "batch_task_slow_poll_duration",
+                "The total duration (s) of slow polls.",
+            ),
+            &task_labels[..],
+        )
+        .unwrap();
+        descs.extend(task_slow_poll_duration.desc().into_iter().cloned());
+
+        let metrics = Self {
+            descs,
+            task_first_poll_delay,
+            task_fast_poll_duration,
+            task_idle_duration,
+            task_poll_duration,
+            task_scheduled_duration,
+            task_slow_poll_duration,
+            delete_task: Arc::new(Mutex::new(Vec::new())),
+        };
+        registry.register(Box::new(metrics.clone())).unwrap();
+        metrics
+    }
+
+    /// Create a new `BatchTaskMetrics` instance used in tests or other places.
+    pub fn for_test() -> Self {
+        Self::new(prometheus::Registry::new())
+    }
+
+    fn clean_metrics(&self) {
+        let delete_task: Vec<TaskId> = if let Ok(mut delete_task) = self.delete_task.try_lock() {
+            if delete_task.is_empty() {
+                return;
+            }
+            std::mem::take(delete_task.as_mut())
+        } else {
+            return;
+        };
+        for id in &delete_task {
+            let stage_id = id.stage_id.to_string();
+            let task_id = id.task_id.to_string();
+            let labels = vec![id.query_id.as_str(), stage_id.as_str(), task_id.as_str()];
+
+            macro_rules! remove {
+                        ($({ $metric:ident, $type:ty},)*) => {
+                            $(
+                                self.$metric.remove_label_values(&labels).unwrap();
+                            )*
+                        };
+                    }
+            for_all_task_metrics!(remove);
+        }
+    }
+
+    pub async fn add_delete_task(&self, id: TaskId) {
+        self.delete_task.lock().await.push(id);
+    }
```

Comment:
Can we use std::Mutex(parking_lot::Mutex) instead of tokio::sync::Mutex? If we use std::Mutex, the lock() may block whole runtime I think.🤔

If we must use tokio::sync::Mutex, lock() for it is a async function and we can't use it in drop.😭

Reply:
AFAIK sync mutex may block the runtime when its guard is held across an `await`, which is not the case here. How will it block the runtime?

Reply:
My bad, you are right.